src/GridFieldMeatballMenuComponent.php

```diff
@@ -0,0 +1,311 @@
+            // a link with a hash in it to the page we're on _now_ #anchor, as opposed to
+            // e.g link/set/here#anchor
+            return $hash && false ? "$link#$hash" : $link;
+        };
+
+        $this->addRootTabActions($record, $linkCallback);
+        $this->addVersionedActions($record, $linkCallback);
+
+        return $this;
+    }
+
+    /**
+     * Add each of the "Root" tabs to the actions for this component
+     *
+     * We expect that a tabbed list of fields will always have a singular root.
+     *
+     * @param DataObject $record
+     * @param callable $linkCallback
+     */
+    protected function addRootTabActions(DataObject $record, callable $linkCallback)
+    {
+        $tabSet = $record->getCMSFields()->first();
+        if (!($tabSet instanceof TabSet)) {
+            return;
+        }
+
+        $first = true;
+        foreach ($tabSet->Tabs() as $tab) {
+            // Skip the first tab if we've opted to
+            if ($first && !$this->getShowFirstTab()) {
+                $first = false;
+                continue;
+            }
+
+            /** @var Tab $tab */
+            $tabID = ($first) ? null : $tab->ID();
+            $this->addActionToGroup([
+                'Title' => $tab->Title(),
+                'Link' => $linkCallback('edit', $tabID),
+                'Type' => 'link'
+            ], 'rootlinks');
+            $first = false;
+        }
+    }
+
+    /**
+     * If the object is versioned (has the {@link Versioned} extension applied) then add
+     * actions to publish/unpublish etc
+     *
+     * @param DataObject $record
+     * @param callable $linkCallback
+     */
+    protected function addVersionedActions(DataObject $record, $linkCallback)
+    {
+        if (!$record->hasExtension(Versioned::class)) {
+            return;
+        }
+
+        if (!$record->latestPublished()) {
+            $this->addActionToGroup([
+                'Title' => _t(__CLASS__ . '.Publish', 'Publish'),
+                'Link' => $linkCallback('publish'),
+                'Type' => 'versioning'
+            ], 'versioned');
+        }
+
+        if ($record->isPublished()) {
+            $this->addActionToGroup([
+                'Title' => _t(__CLASS__ . '.Unpublish', 'Unpublish'),
+                'Link' => $linkCallback('unpublish'),
+                'Type' => 'versioning'
+            ], 'versioned');
+        }
+
+        $this->addActionToGroup([
+            'Title' => _t(__CLASS__ . '.Delete', 'Delete'),
+            'Link' => $linkCallback('archive'),
+            'Type' => 'versioning'
+        ], 'versioned');
+    }
+
+    public function getColumnContent($gridField, $record, $columnName)
+    {
+        $this->buildDefaultActions($gridField, $record);
+
+        $templateData = ArrayData::create([
+            'Actions' => Convert::raw2json(array_values($this->getActions())),
+        ]);
+
+        return $templateData->renderWith(static::class);
```

Comment:
I know we made this change together, but this would mean child classes would have to define their own template, where they might just want to modify some behaviour. Perhaps we should stick to `__CLASS__` (not late bound) and let child classes overload this method if they want to, or projects could just modify the template in its own namespace in their theme

Reply:
Remember originally as we made the change it went to find a template given the class - reverting to this original state will allow for template hierarchy to be accounted into this selection, and child classes will fall back to this parent class' template.

Perhaps this would be better, allowing both flexibility and the static short hand.